doc/macsec/MACsec_hld.md

```diff
@@ -0,0 +1,613 @@
+|    delete_transmit_sa    | DEL APP_DB[MACSEC_SA_EGRESS]<br>WAIT DEL STATE_DB[MACSEC_EGRESS_SA]                                       |                                            |
+|    enable_transmit_sa    | SET APP_DB[MACSEC_SC:ENCODING_AN]=PARAM<br>WAIT SET STATE_DB[MACSEC_EGRESS_SA]                            |                                            |
+|   disable_transmit_sa    |                                                                                                           |                                            |
+
+***WAIT : To subscribe the target table and to use the select operation to query the expected message***
+
+#### 3.4.4 MACsec Orch
+
+The MACsecOrch is introduced in the Orchagent to handle configuration requests. It monitors MACsec related tables in APP DB and convert those messages to SAI commands to manage the MACsec object. All MACsec SAI APIs are defined at (<https://github.com/opencomputeproject/SAI/blob/master/inc/saimacsec.h>).
+
+##### 3.4.4.1 Functions
+
+The following are all functions that MACsec Orch need to implement.
+
+- Enable MACsec
+  1. Monitor the SET message from the MACsec Port Table in APP DB
+  2. Create ingress/egress MACsec ports
+  3. Create ingress/egress MACsec ACL tables
+  4. Create an ACL entry to permit EAPOL packets to 802.1x destination MAC address
+  5. Create an ACL entry to drop packets (to be later used for macsec_flow), If PROTECT_ENABLE. Otherwise, not drop
+  6. Bind the ingress/egress ACL tables to the ingress/egress MACsec ports
+  7. Set Flex counters of MACsec port stats
+  8. Set State DB
+
+- Disable MACsec
+  1. Monitor the DEL message from  the MACsec Port Table in APP DB
+  2. Delete Flex counter of MACsec port stats
+  3. Unbind the ingress/egress ACL tables from the ingress/egress MACsec ports
+  4. Delete the ingress/egress ACL tables
+  5. Delete the ingress/egress MACsec ports
+  6. Del State DB
+
+- Create SC
+  1. Monitor the SET message from the MACsec Ingress/Egress SC Table in APP DB
+  2. Create ingress/egress MACsec flow
+  3. Create ingress/egress MACsec SC
+  4. Set Flex counter of MACsec SC and flow stats
+
+- Delete SC
+  1. Monitor the DEL message from the MACsec Ingress/Egress SC Table in APP DB
+  2. Collect MACsec SC and flow stats
+  3. Delete Flex counter of MACsec SC
+  4. Delete the ingress/egress MACsec SC
+  5. Delete the ingress/egress MACsec SA
+
+- Create Ingress SA
+  1. Monitor the DEL message from the MACsec  Ingress SA Table
+  2. Collect SA Stats
+  3. Delete Flex counter of MACsec ingress SA stats
+  4. Delete ingress MACsec SA
+
+- Create Egress SA
+  1. Monitor the SET message from the MACsec Egress SC Table, only if ENCODING_AN change
+  2. The ACTIVE filed should be true, otherwise exit this process
+  3. Create egress MACsec egress SA
+  4. Set Flex Counter of MACsec SA stats
+
+- Delete SA
+  1. Monitor the DEL message from the MACsec SA Table
+  2. Collect SA Stats
```

Comment:
Is this intended in Delete SA?

Reply:
Yes, we need to finally collect SA stats from counter DB when we delete it.